### ISSUE TITLE
Cannot read property 'data' of null errors

### ISSUE THREAD

**Issue by coderangerdan** [stevenbenner/jquery-powertip#165]
I am getting a lot of errors being reported by Application Insights (exception tracker like BugSnag) daily, and its taken me a while to track down the line (as its all minified and bundled) and its this line:

https://github.com/stevenbenner/jquery-powertip/blob/3c1d8434a5a53233b1d6d03954edd9094deeb9af/src/tooltipcontroller.js#L379

The errors is "Cannot read property 'data' of null" so I think there needs to be some sort of guard checking around this part of code; unfortunately i cant reproduce it myself so as to find out why it may be null.

**Comment by stevenbenner** [comment]
Interesting. It's not obvious from looking at the code how it could get in this state. The code that sets the `activeHover` property is pretty straightforward:

```javascript
function showTip(element) {
	/* ... */
	session.activeHover = element;
	/* ... */
}

function hideTip(element) {
	/* ... */
	session.desyncTimeout = clearInterval(session.desyncTimeout);
	/* ... */
	tipElement.fadeOut(options.fadeOutTime, function fadeOutCallback() {
		/* ... */
		session.activeHover = null;
		/* ... */
	});
	/* ... */
}
```
The only way I can see this error happening is if the `activeHover` property is somehow deleted early, or the `closeDesyncedTip()` function is run after the tooltip fadeout has completed.

I'm hesitant to implement a guard fix without knowing what is happening because several important behaviors depend on this value being accurate. If it's not then it may cause weird bugs.

Is there any info that you have that might help track down what is happening?

* Does this affect only specific browsers?
* Does this happen only on specific pages?
* How are you using PowerTip? (just a simple `$(foo).powerTip()` on load? or something more complicated?)
* What selector are you using to run PowerTip (is it ever operating on a document fragment?)
* Do you ever set the `fadeOutTime` option?
* Do you ever use the API to `hide()` or `toggle()` tooltips manually?
* Do you ever use the API to `destroy()` tooltips?
* Are the page especially dynamic? (do the elements ever get hidden or deleted?)

**Comment by Trycoder** [comment]
I also face the same issue frequently when a grid cell is randomly hovered.
i guess some sort of solution to be put in overhere : line no 935 of powertip.js in function hideTip(element) :element.data(DATA_HASACTIVEHOVER, false);

**Comment by stevenbenner** [comment]
@Trycoder can you answer any of the questions from my previous comment? The information would be really helpful to try to figure out what is causing the issue.

> * Does this affect only specific browsers?
> * Does this happen only on specific pages?
> * How are you using PowerTip? (just a simple `$(foo).powerTip()` on load? or something more complicated?)
> * What selector are you using to run PowerTip (is it ever operating on a document fragment?)
> * Do you ever set the `fadeOutTime` option?
> * Do you ever use the API to `hide()` or `toggle()` tooltips manually?
> * Do you ever use the API to `destroy()` tooltips?
> * Are the page especially dynamic? (do the elements ever get hidden or deleted?)

Also, it sounds like you can get this error to happen regularly. If you can provide an example (jsfiddle, html file, or something) that would make tracking it down much easier.

**Comment by coderangerdan** [comment]
Unfortunately it’s hard to track down, I just see it in my application insights logs, I can’t reproduce it myself yet.

> does this affect only specific browsers
Not that I can see

> does this happen on specific pages
Not that I can see, but call stack doesn’t help

> how are you using power tip
```
$(‘A.tooltip’).powerTip( { options... } );

$(‘A.tooltip).on( {
  powerTipPreRender: function() {
    $( this ).data( ‘powertip’, getContentForTooltip() );
  },
  powerTipRender: function() {
    $( ‘DIV#toolTip.heading’ ).attr( ‘title’, ‘some changes’ );
  }
);
```

> do I use fadeTimeout
Nope

> do I use api to toggle or hide
Yes, I use $.powerTip.hide();

>do I use api to destroy
Yes, I do use $.powerTip.destroy() when the main elements on the page need refreshing (a change of navigation) due to an issue I have reported with tooltips taking exponentially longer to initialise on my pages and the destroy fixes that (sorry still not got around to creating a fragment example)

> are the pages dynamic
Yes elements get replaced (therefore deleted) often

**Comment by stevenbenner** [comment]
I've been trying to find some situation where I can reproduce the error, but so far I haven't found any. So I have a couple more follow-up questions:

* What version of jQuery are you using?
* What version of PowerTip are you using?
* How do you delete elements on the page? (jQuery `$.remove()`, DOM `removeChild`, DOM `remove()`, other?)
* When replacing elements, do you remove the elements with tooltips individually, or do you remove some higher level container that holds the elements with tooltips?

I have a hunch that this is related to the usage of the `destroy()` method. I've discovered a bug that if you destroy a tooltip while it is currently open it doesn't trigger the normal tooltip `hide()` code, leaving the desync checker running indefinitely.

However, this doesn't explain the problem. The `session.activeHover.data()` line that is generating the error actually is guarded by `if (session.isTipOpen ...`. There is no point in the tooltip life-cycle where `isTipOpen` is `true` and the `activeHover` property is `null` (well, at least according to the code).

**Comment by stevenbenner** [comment]
The most likely case that would generate this error should be fixed with commit f4677633331d2767a694deb5a1b5ac198dacedbd.

Note that this probably doesn't fix whatever is actually causing the state to get corrupted on your site. This change simply makes sure that the API `destroy()` calls will now force any open tooltips to close, giving them an opportunity to cancel their desync checking. Whatever is going wrong is likely causing problematic tooltip behavior, and the fix I implemented will not correct that unknown issue for the affected users.

Since there has been no response to my latest questions for a while I cannot do any further investigation, so I'm closing this issue.

If you ever get a chance, please review the questions I asked in my last comment.

**Comment by MikeAllen775** [comment]
@stevenbenner Sorry to resurrect this, but I'm able to recreate the same issue reliably, but at another code location.  On line 912: hideTip(session.activeHover), where the activeHover property is null.  Although it most likely has a lot to do with the way I'm loading the content, which is:  On attaching the powertip, I display "loading" and then use jQuery.getJSON, update the powertipjs data property, and also refresh the currently showing powerTip (as long as the unique content's div is still in the dom).  I do this all by attaching powerTip on a mouseover event on the element (because I need to account for dynamically added elements).

I'm not sure how much of the above is necessary to recreate the problem.  With the above actions, if I alternate rapidly between multiple powerTip elements (I have around 20 in a table column, and just run up and down the list), occasionally the error occurs.  An additional oddity is that whenever the error occurs, the session.isTipOpen value is false on line 910, which should have precluded it even getting to line 912.

I know you can't guard against every scenario, but I've found that just adding an "if (session.activeHover !== null)" check, the problem never occurs, and everything else behaves normally.  

By the way, I love this plugin!  Great work!
